lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go

```diff
@@ -89,13 +94,73 @@ func (s *subscriberdbServicer) ListSubscribers(ctx context.Context, req *lte_pro
 		subProto.NetworkId = &protos.NetworkID{Id: networkID}
 		subProtos = append(subProtos, subProto)
 	}
+
+	// Short-circuit subscriber data transmission by comparing cloud & gateway digests.
+	// If digest generation fails, the error is swallowed to not affect the main functionality.
+	//
+	// Note: currently the logic only applies to a request for the first page to avoid concurrency issues;
+	// more to be implemented with gateway subscriber cache that will canonize cloud digests.
+	noUpdates, digest := false, &req.PreviousDigest
+	if req.PageToken == "" {
+		digest, err = getDigest(gateway, apnsByName, apnResourcesByAPN)
+		if err != nil {
+			glog.Errorf("Generating digest for subscribers in network of gateway %s failed", gatewayID)
+		} else {
+			noUpdates = req.PreviousDigest == *digest
+			if noUpdates {
+				subProtos = []*lte_protos.SubscriberData{}
+			}
+		}
+	}
+
 	listRes := &lte_protos.ListSubscribersResponse{
 		Subscribers:   subProtos,
 		NextPageToken: nextToken,
+		Digest:        *digest,
+		NoUpdates:     noUpdates,
 	}
 	return listRes, nil
 }
 
+// getDigest loads all subscribers registered on the current network, and returns
+// a deterministic subscriber flat digest.
+func getDigest(gateway *protos.Identity_Gateway, apnsByName map[string]*lte_models.ApnConfiguration, apnResourcesByAPN lte_models.ApnResources) (*string, error) {
+	networkID := gateway.NetworkId
+	lc := configurator.EntityLoadCriteria{
+		PageSize:           0,
+		PageToken:          "",
+		LoadConfig:         true,
+		LoadAssocsToThis:   true,
+		LoadAssocsFromThis: true,
+	}
+	subEnts, _, err := configurator.LoadAllEntitiesOfType(
+		networkID, lte.SubscriberEntityType, lc, serdes.Entity,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "load all subscribers in network of gateway %s", networkID)
+	}
+
+	subProtosById := map[string]proto.Message{}
+	for _, sub := range subEnts {
+		subProto, err := convertSubEntsToProtos(sub, apnsByName, apnResourcesByAPN)
+		if err != nil {
+			return nil, err
+		}
+		subProto.NetworkId = &protos.NetworkID{Id: networkID}
+
+		index := subProto.Sid.Id
+		subProtosById[index] = subProto
+	}
+
+	digest, err := mproto.MarshalManyDeterministic(subProtosById)
+	if err != nil {
+		return nil, err
+	}
+	digestB64 := b64.StdEncoding.EncodeToString(digest)
```

Comment:
I wonder if you even need an extra overhead of base64 encoding & extra memory allocation instead of just declaring & passing digests as []byte?
Just change 'string digest = 3;' to 'bytes digest = 3;' in your proto definition?

Reply:
Thanks for the suggestion @emakeev, I'm currently using `string` simply because it appeared to be more intuitive to me, but I'll try refactoring it into `[]byte` & get you posted on this!

Reply:
The digest is meant to be ~4 bytes, and we're just getting/setting, not building. So using string is probably easier. (Though if we were sending something longer, or building the digest, bytes probably makes more sense.)

Ref for how we do/did it for mconfigs https://github.com/magma/magma/blob/509e5b4bcdb68b0021044fb1dc8cf47554446b8f/orc8r/protos/mconfig.proto#L34-L45